src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,412 @@
+                timestamp is outside our bounds
+
+        Raises:
+            IndexError: when requesting a timestamp outside the range this container holds
+
+        Returns:
+            index where the value for the given timestamp can be found.
+        """
+        timestamp = self.normalize_timestamp(timestamp)
+
+        if not allow_outside_range and (
+            self._datetime_newest + timedelta(seconds=1 * self._sampling_period_s)
+            < timestamp
+            or timestamp < self._datetime_oldest
+        ):
+            raise IndexError(
+                f"Requested timestamp {timestamp} is is "
+                f"outside the range [{self._datetime_oldest} - {self._datetime_newest}]"
+            )
+
+        return self._wrap(
+            round(
+                abs((self._time_index_alignment - timestamp).total_seconds())
+                / self._sampling_period_s
+            )
+        )
+
+    def window(
+        self, start: datetime, end: datetime, force_copy: bool = False
+    ) -> Container:
+        """Request a view on the data between start timestamp and end timestamp.
+
+        If the data is not used immediately it could be overwritten.
+        Always request a copy if you keep the data around for longer.
+
+        Will return a copy in the following cases:
+        * The requested time period is crossing the start/end of the buffer.
+        * The requested time period contains missing entries.
+        * The force_copy parameter was set to True (default False).
+
+        This means, if the caller needs to modify the data to account for
+        missing entries, they can safely do so.
+
+        Args:
+            start: start time of the window.
+            end: end time of the window.
+            force_copy: optional, default False. If True, will always create a
+                copy of the data.
+
+        Returns:
+            the requested window
+        """
+        assert start < end
+
+        start_index = self.datetime_to_index(start)
+        end_index = self.datetime_to_index(end)
+
+        # Requested window wraps around the ends
+        if start_index >= end_index:
+            window = self._buffer[start_index:]
+
+            if end_index > 0:
+                if isinstance(self._buffer, list):
+                    window += self._buffer[0:end_index]
+                else:
+                    window = np.concatenate((window, self._buffer[0:end_index]))
+            return window
+
+        def in_window(window) -> bool:
+            return self._in_window(window, start) or self._in_window(window, end)
+
+        # Return a copy if there are none-values in the data
+        if force_copy or any(map(in_window, self._missing_windows)):
+            return deepcopy(self._buffer[start_index:end_index])
+
+        return self._buffer[start_index:end_index]
+
+    def is_missing(self, timestamp: datetime) -> bool:
```

Comment:
is this part of the public api?

Reply:
IMO that can be private.

Reply:
I thought it is rather useful, but sure, I can make it private

Reply:
This should be public, it can be used to check if a specific sample datetime is already written or not.